### PR TITLE
Added new field policy_id to identify policies, modified rua, and policy 'a' field

### PR DESCRIPTION
**Pull request by prbinu**
And few minor additions/changes. I have removed few lines. If you think those lines are required, please rephrase it

### CODE REVIEW

File: spec.md [hunk 1]
```diff
@@ -204,15 +205,15 @@ to DANE:
   recipient domain.
 
 * Security: DANE offers an advantage against policy-lookup DoS attacks; that is,
-  while a DNSSEC-signed NX response to a DANE lookup authoritatively indicates
+  while a DNSSEC-signed MX response to a DANE lookup authoritatively indicates
```
Comment (danmarg) [inline spec.md]: This was meant to be "NXDOMAIN", not "MX." In other words, a DNSSEC-signed "non-existent" response is authoritative. 

Probably should say "NXDOMAIN", though. :)

File: spec.md [hunk 2]
```diff
@@ -246,8 +249,9 @@ Policies MUST specify the following fields:
   clients SHOULD cache a policy for up to this value from last policy fetch
   time.
 * rua: Address to which aggregate feedback MAY be sent (comma-separated
-  plain-text list of email addresses, optional). For example,
-  "mailto:postmaster@example.com" from [@!RFC3986].
+  plain-text list of email addresses or an HTTPS endpoint, optional). For example,
+  "mailto:postmaster@example.com" from [@!RFC3986] or "https://example.com/sts-report".
+* id: A unique identifier to identify a policy. It could be first 16 bytes of SHA256(all fields except 'id') hash
```
Comment (danmarg) [inline spec.md]: Good idea. This will help process reports or debug policy fetching.

File: spec.md [hunk 3]
```diff
@@ -246,8 +249,9 @@ Policies MUST specify the following fields:
   clients SHOULD cache a policy for up to this value from last policy fetch
   time.
 * rua: Address to which aggregate feedback MAY be sent (comma-separated
-  plain-text list of email addresses, optional). For example,
-  "mailto:postmaster@example.com" from [@!RFC3986].
+  plain-text list of email addresses or an HTTPS endpoint, optional). For example,
```
Comment (danmarg) [inline spec.md]: Regarding reporting via HTTP(S): I think we may want this, but the spec doesn't currently say what the protocol for reporting looks like. Is it to just POST to this URI? If so, we should say that.

File: spec.md [hunk 4]
```diff
@@ -266,14 +270,15 @@ The formal definition of the SMTP STS format, using [@!RFC5234], is as follows:
...
                        ; components other than sts-version and
-                       ; sts-to may appear in any order
+                       ; sts-m may appear in any order
 
     sts-version     = "v" *WSP "=" *WSP %x53 %x54 %x53 %x31
 
     sts-sep         = *WSP %x3b *WSP
 
-    sts-to          = "to" *WSP "=" *WSP ( "true" / "false" )
+    sts-m           = "to" *WSP "=" *WSP ( "enforce" / "report" )
```
Comment (danmarg) [inline spec.md]: @mrisher I think you had some similar thoughts. What do you think?

File: spec.md [hunk 5]
```diff
@@ -319,43 +327,43 @@ recursive resolver. Consequently, a sender MAY treat a policy as valid for up to
...
-mechanism specified by the existing cached policy.
-
-Note, however, as described in detail in _Policy_ _Application_, that new
-policies MUST NOT be considered as valid if they do not validate on first
-application. That is, a freshly fetched (and unused) policy that has not
-successfully been applied MUST be disregarded.
+When fetching a new policy, or when fetching a policy for a domain with an 
+expired policy or if the sender detects a policy update (by checking policy id
+or computing SHA256 of policy), the new policy SHOULD be authenticated before use. 
+If the policy domain is DNSSEC enabled, then sender MAY skip explicit policy
```
Comment (danmarg) [inline spec.md]: If the policy is served over DNSSEC but the policy specifies a=webpki, do we not still want to do webpki authentication? After all, the policy requested it. 

This seems like it goes more to Alex's point that we should allow an OR statement in the auth mechanism--e.g. "dnssec,webpki" where either is acceptable. What do you think, Binu?

File: spec.md [hunk 6]
```diff
@@ -547,7 +555,7 @@ In addition, the authors leave currently open the following details:
 * Whether and how more detailed "forensic reporting" should be accomplished, as
   discussed in the section _Failure_ _Reporting_.
 
-* The registration of the .well-known/smtp-sts URI as per [@!RFC5785].
+* The registration of the .well-known policy._smtp-sts URI prefix as per [@!RFC5785].
```
Comment (danmarg) [inline spec.md]: To be a nitpicker, we don't need to register the *prefix*: RFC 5785 concerns only .well-known suffix registration. So the hostname on the policy domain is up to us and requires no registration. ;)

File: spec.md (hunk 4 above)
Comment (prbinu) [inline spec.md]: sts-to is confusing

File: spec.md (hunk 5 above)
Comment (prbinu) [inline spec.md]: If the policy is served over DNSSEC, then there is no need to further authenticate the policy. Yes, it should be either dnssec OR webpki.

File: spec.md (hunk 4 above)
Comment (mrisher) [inline spec.md]: I agree, and on top of that, it looks like I jumbled up the definition of "to" in one of the Friday edits. I like "mode" better.

File: spec.md [hunk 7]
```diff
@@ -327,6 +327,17 @@ recursive resolver. Consequently, a sender MAY treat a policy as valid for up to
 {expiration time} + {DNS TTL}. Publishers SHOULD thus continue to expect senders
 to apply old policies for up to this duration.
 
+### Policy Updation
+If the policy is authenticated over webpki, then in order to change an STS policy,
+the policy owner has to update the policy in two places - DNS TXT RR and HTTPS
+endpoint. To address a potential race-condition state, or in other words if the
+policy update in HTTPS lags behind the DNS TXT record or vice versa, the policy
+fetched during that period will fail to authenticate (and thus invalid). In such
```
Comment (mrisher) [inline spec.md]: "and *is* thus invalid"

File: spec.md [hunk 8]
```diff
@@ -223,7 +223,7 @@ distribution. See the section _Future_ _Work_ for more discussion.)
 Policies MUST specify the following fields:
 
 * v: Version (plain-text, required). Currently only "STS1" is supported.
-* m:  (plain-text, required). If "enforce", the receiving MTA requests
+* m: Mode (plain-text, required). If "enforce", the receiving MTA requests
   that messages be delivered only if they conform to the STS policy. If "report"
   the receiving MTA requests that failure reports be delivered, as specified by
   the `rua` parameter.
```
Comment (mrisher) [inline spec.md]: We should note that reports are to be delivered in *both* cases, right?

File: spec.md [hunk 9]
```diff
@@ -358,10 +358,12 @@ for policy authentication:
...
 
 When fetching a new policy, or when fetching a policy for a domain with an 
-expired policy or if the sender detects a policy update (by checking policy id
-or computing SHA256 of policy), the new policy SHOULD be authenticated before use. 
+expired policy or if the sender detects a policy update (by checking policy id),
+the new policy SHOULD be authenticated before use. 
 If the policy domain is DNSSEC enabled, then sender MAY skip explicit policy
-authentication mechanisms mentioned above ('a' field in the policy).
+authentication mechanisms mentioned above ('a' field in the policy). The 
+requirement is that the policy should be authenticated by either dnssec OR webpki
```
Comment (mrisher) [inline spec.md]: please capitalize `DNSSEC`

File: spec.md [hunk 10]
```diff
@@ -250,7 +250,7 @@ Policies MUST specify the following fields:
   time.
 * rua: Address to which aggregate feedback MAY be sent (comma-separated
   plain-text list of email addresses or an HTTPS endpoint, optional). For example,
-  "mailto:postmaster@example.com" from [@!RFC3986] or "https://example.com/sts-report".
+  "mailto:postmaster@example.com" from [@!RFC3986] or HTTP POST to "https://example.com/sts-report".
```
Comment (danmarg) [inline spec.md]: Does it make me a weenie if I request an 80 char line wrap? ;)

File: spec.md [hunk 11]
```diff
@@ -327,6 +327,17 @@ recursive resolver. Consequently, a sender MAY treat a policy as valid for up to
 {expiration time} + {DNS TTL}. Publishers SHOULD thus continue to expect senders
 to apply old policies for up to this duration.
 
+### Policy Updation
```
Comment (danmarg) [inline spec.md]: s/Updation/Updating/

File: spec.md [hunk 12]
```diff
@@ -118,6 +118,7 @@ of the delivery domain) can perform such a downgrade or interception attack.
 This document defines a mechanism for recipient domains to publish policies
 specifying:
 
+   * MTA that handles emails for the recipient domain
```
Comment (danmarg) [inline spec.md]: I wouldn't bother to call this out (the "MTA that handles email for the recipient domain") because that's really what we mean by the third bullet point (how to validate the TLS server certificate). What do you think?

(Note I'm making a few edits to your pull request in another branch, so just commenting here so you're aware of my reasoning in case you disagree.)

File: spec.md [hunk 13]
```diff
@@ -204,15 +205,15 @@ to DANE:
...
 * Security: DANE offers an advantage against policy-lookup DoS attacks; that is,
-  while a DNSSEC-signed NX response to a DANE lookup authoritatively indicates
+  while a DNSSEC-signed NXDOMAIN response to a DANE lookup authoritatively indicates
   the lack of a DANE record, such an option to authenticate policy non-existence
   does not exist when looking up a policy over plain DNS.
 
 # Policy Semantics
 
-SMTP STS policies are distributed at the Policy Domain either through a new
+SMTP STS policies are distributed from Policy Domain's DNS either through a new
```
Comment (danmarg) [inline spec.md]: Rephrased this slightly. 

(Note I'm making a few edits to your pull request in another branch, so just commenting here so you're aware of my reasoning in case you disagree.)

File: spec.md [hunk 14]
```diff
@@ -232,12 +233,14 @@ Policies MUST specify the following fields:
   for this domain might be handled by any MX whose hostname is a subdomain of
   "example.com" or "example.net."
 * a: The mechanism to use to authenticate this policy itself. See the section
-  _Policy_ _Authentication_ for more details. Possible values are:
-  * webpki:URI, where URI points to an HTTPS resource at the recipient domain
-    that serves the same policy text.
+  _Policy_ _Discovery_ _&_ _Authentication_ for more details. Possible values are:
+  * webpki: Indicating that the policy is expected to be served over HTTPS on a
```
Comment (danmarg) [inline spec.md]: Why this change? I don't think it matters much one way or another, but this is a slightly larger change that should maybe be called out (or made its own pull request). 

We also should not call this a "well-known" URI if we don't plan to use .well-known (RFC 5785), which means we can remove that consideration from the IANA considerations section, right?

File: spec.md [hunk 15]
```diff
@@ -232,12 +233,14 @@ Policies MUST specify the following fields:
...
+    well-known URI: https://policy._smtp_sts.example.com/current. The 
+    example.com is the recipient's domain, and the endpoint serves the same policy
+    text.
   * dnssec: Indicating that the policy is expected to be served over DNSSEC.
 * c: Constraints on the recipient MX's TLS certificate (plain-text, required).
-  See the section _Policy_ _Validation_ for more details. Possible values are:
+  See the section _Recipient's_ _Certificate_ _Validation_ for more details. Possible values are:
```
Comment (danmarg) [inline spec.md]: I prefer "Policy Validation" since it matches nicely with "Policy Authentication." I.e. first you authenticate a policy, then you validate *against* a policy, then you...handle failures, report, etc. No?

File: spec.md [hunk 16]
```diff
@@ -246,8 +249,9 @@ Policies MUST specify the following fields:
   clients SHOULD cache a policy for up to this value from last policy fetch
   time.
 * rua: Address to which aggregate feedback MAY be sent (comma-separated
-  plain-text list of email addresses, optional). For example,
-  "mailto:postmaster@example.com" from [@!RFC3986].
+  plain-text list of email addresses or an HTTPS endpoint, optional). For example,
+  "mailto:postmaster@example.com" from [@!RFC3986] or HTTP POST to "https://example.com/sts-report".
+* id: A unique identifier to identify a policy. It could be first 16 bytes of SHA256(all fields except 'id') hash
```
Comment (danmarg) [inline spec.md]: One comment about this: Since we already specified that reports contain the raw string of the applied policy, is there utility in having a policy *id*? It seems to achieve the same thing, but with a layer of indirection (which may be good or bad; I'm not sure).

File: spec.md [hunk 17]
```diff
@@ -246,8 +249,9 @@ Policies MUST specify the following fields:
   clients SHOULD cache a policy for up to this value from last policy fetch
   time.
 * rua: Address to which aggregate feedback MAY be sent (comma-separated
-  plain-text list of email addresses, optional). For example,
-  "mailto:postmaster@example.com" from [@!RFC3986].
+  plain-text list of email addresses or an HTTPS endpoint, optional). For example,
+  "mailto:postmaster@example.com" from [@!RFC3986] or HTTP POST to "https://example.com/sts-report".
```
Comment (danmarg) [inline spec.md]: I'm moving the thing about POST'ing as a note in the "Failure Reporting" section.

File: spec.md (hunk 12 above)
Comment (prbinu) [inline spec.md]: I think we need to put this explicitly because this proposed policy actually does two things:
1. Detect MX mismatches
2. Validate certificate.

Validate certificate is applicable only if the MX matches with the one mentioned in the policy. On top of that mentioning the first point also adds more clarity to readers.

File: spec.md (hunk 15 above)
Comment (prbinu) [inline spec.md]: Policy validation is fine, but what we described in that section is only part of the policy validation. Like i mentioned before, cert validation is only one part. The other part is the MX name check. Based on the description, I proposed a heading change to make it more explicit. If you would like to keep 'Policy Validation', make sure we describe all validation steps in that section.

File: spec.md (hunk 16 above)
Comment (prbinu) [inline spec.md]: Otherwise how to we detect a policy change?

Comment (danmarg) [inline spec.md]: In which case? I'm not sure you ever *need* to detect a policy change, but you can always diff it against the old cached policy. Since they're small and we rely on senders storing some long-lived cache of policy contents anyway, I still don't see the need for an ID. What am I missing?

File: spec.md (hunk 15 above)
Comment (danmarg) [inline spec.md]: Good point. Agreed on this. 

I was only bike-shedding about "authentication" and "validation" because we had that long talk about the different stages/semantics at MAAWG, so I was trying to hew to that terminology. ;)

File: spec.md [hunk 18]
```diff
@@ -319,43 +327,56 @@ recursive resolver. Consequently, a sender MAY treat a policy as valid for up to
...
-policies MUST NOT be considered as valid if they do not validate on first
-application. That is, a freshly fetched (and unused) policy that has not
-successfully been applied MUST be disregarded.
+When fetching a new policy, or when fetching a policy for a domain with an 
+expired policy or if the sender detects a policy update (by checking policy id),
+the new policy SHOULD be authenticated before use. 
+If the policy domain is DNSSEC enabled, then sender MAY skip explicit policy
+authentication mechanisms mentioned above ('a' field in the policy). The 
+requirement is that the policy should be authenticated by either dnssec OR webpki
+mechanism.
```
Comment (danmarg) [inline spec.md]: I'm a little unsure about this last part. If a=webpki, should senders nonetheless skip webpki authentication if  DNSSEC is used? I would say no; the policy domain specified to use webpki. This is a place where an explicit "OR" would be nice, though.

File: spec.md [hunk 19]
```diff
@@ -295,6 +300,9 @@ The formal definition of the SMTP STS format, using [@!RFC5234], is as follows:
     sts-auri        = "rua" *WSP "=" *WSP
                        sts-uri *(*WSP "," *WSP sts-uri)
 
+    sts-id           = "id" *WSP "=" *WSP 1*16HEXDIG
```
Comment (danmarg) [inline spec.md]: I think I said in email, but "ID" seems unnecessary to me since "applied policy" is a field in the report. 

That said, see Viktor's comments on the UTA mailing list about a DNS nonce, etc.

File: spec.md (hunk 12 above)
Comment (danmarg) [inline spec.md]: I see. I'll add something back in and you can comment. :)
